File: lex.js
function lex(string) {
	var tokens = [];

	for (var c = 0; c < string.length; c++) {
		// Skip nothing
		if (" \n	".includes(string[c])) {
			continue;
		}

		tokens.push({value: "", type: ""});
		var current = tokens[tokens.length - 1];

		if (isAlpha(string[c])) {
			current.type = "text";
			while (isAlpha(string[c])) {
				current.value += string[c];
				c++;
			}

			// Quit on detection of label
			if (string[c] == ":") {
				current.type = "label";
				return tokens;
			}

			// Test for selector "asd[1]"
			if (string[c] == "[") {
				current.selector = "";
				c++;
				while (string[c] != "]") {
					current.selector += string[c];
					c++;
				}
			}
		} else if (string[c] == ";") {
			while (c < string.length) {
				c++;
			}

			// If we have not done any other tokens
			if (tokens.length == 1) {
				return [];
			}
		} else if (isNum(string[c])) {
			current.type = "int";
			while (isNum(string[c])) {
				current.value += string[c];
				c++;
			}

			c--;
		} else if (string[c] == "'") {
			current.type = "char";
			c++;
			current.value = string[c];
			c += 2;

		} else if (string[c] == "\"") {
			current.type = "string";
			c++;
			while (string[c] != "\"") {
				current.value += string[c];
				c++;
			}
		}
	}

	return tokens;
}

function isAlpha(char) {
	if ("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_".includes(char)) {
		return true;
	} else {
		return false;
	}
}

function isNum(char) {
	if ("0123456789".includes(char)) {
		return true;
	} else {
		return false;
	}
}

File: cli/assemble/lex.c
#include <string.h>
#include <stdbool.h>
#include "object.h"

int isAlpha(char c) {
	if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c == '.' || c == '_')) {
		return 1;
	} else {
		return 0;
	}
}

int isDigit(char c) {
	if (c >= '0' && c <= '9') {
		return 1;
	} else {
		return 0;
	}
}

// Locate var, arr, lbl, from objects in memory from type
int locateObject(struct Memory *memory, char *name, int type) {
	for (int i = 0; i < memory->length; i++) {
		// Make sure to check type first. Name can sometimes be unitialized.
		if (memory->d[i].type == type && !strcmp(memory->d[i].name, name)) {
			return i;
		}
	}

	return -1;
}

// Lex single line, then quit
int lex(struct Token *tokens, char *line) {
	int c = 0;
	int token = 0;
	while (line[c] != '\0') {
		// Skip chars
		while (line[c] == ' ' || line[c] == '\t') {
			c++;
		}

		// Skip comments
		if (line[c] == ';') {
			while (line[c] != '\n' || line[c] == '\0') {
				c++;
			}
		}

		// Check if this is a nothing line (comments, blank)
		if (line[c] == '\n' || line[c] == '\0') {
			return token;
		}

		tokens[token].addressOf = 0;

		if (line[c] == '&') {
			tokens[token].addressOf = 1;
			c++;
		}

		tokens[token].length = 0;
		tokens[token].type = 0;
		tokens[token].value = 0;
		if (isAlpha(line[c])) {
			tokens[token].type = TEXT;
			while (isAlpha(line[c])) {
				tokens[token].text[tokens[token].length] = line[c];
				tokens[token].length++;

				c++;
			}

			if (line[c] == ':') {
				tokens[token].type = LABEL;
				c++;
			}
		} else if (isDigit(line[c])) {
			tokens[token].type = DIGIT;
			while (isDigit(line[c])) {
				tokens[token].value *= 10;
				tokens[token].value += (line[c] - '0');
				c++;
			}
		} else if (line[c] == '\'') {
			tokens[token].type = DIGIT;
			c++;
			tokens[token].value = line[c];
			c += 2; // Skip ' and goto next char
		} else if (line[c] == '"') {
			tokens[token].type = STRING;
			c++; // Skip "
			while (line[c] != '"') {
				tokens[token].text[tokens[token].length] = line[c];
				tokens[token].length++;
				c++;
			}

			c++; // Skip "
		}

		// Always null terminate string
		tokens[token].text[tokens[token].length] = '\0';
		token++;
	}

	return token;
}
